[Woo POS][Local Catalog] Refactor and fix pull-to-refresh logic in products and variations

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
WOOMOB-1556

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR refactors pull-to-refresh logic for both products and variations in the WooCommerce POS local catalog feature. The changes consolidate different refresh behaviors by introducing new `refreshProducts()` and `refreshVariations()` methods that delegate to either local catalog incremental sync or remote data source refresh based on feature flags and sync requirements.

### Test Steps
<!-- Describe how to test your changes. Include only what’s needed for the reviewer to validate the behavior.
For new features, outline the main user flow or goal rather than every tap or click — the reviewer should be able to complete the flow naturally.
If applicable, mention key devices, scenarios, or edge cases to verify. -->

* Crucial: Verify that there's no regression to the existing PTR behavior in the hybrid data source.
* Verify that local catalog incremental sync is performed for both variations and products.
* Verify that local catalog incremental sync is not performed if the local catalog is not enabled.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
N/A

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Refactor and fix pull-to-refresh logic in products and variations view models

Key changes:
- A new `refreshProducts` and `refreshVariations` method has been introduced in the `WooPosProductsDataSource` layer to encapsulate the refresh logic.
- The view models now call these new methods, removing the need for feature flag checks and direct interaction with sync repositories.
- The `WooPosProductsInDbDataSource` now handles incremental sync logic when pull-to-refresh is triggered, falling back to a remote fetch if incremental sync isn't applicable.

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/products/WooPosProductsViewModel.kt:
```diff
@@ -337,32 +320,42 @@ class WooPosProductsViewModel @Inject constructor(
+                when (result) {
+                    is WooPosProductsDataSource.ProductsResult.Cached -> {
+                        if (result.products.isNotEmpty()) {
+                            _viewState.value = result.products.toContentState()
```
Cached results are intermediate, meaning we're still waiting for Remote results, so if empty Cached result is received, we should keep showing PTR indicator as we're awaiting Remote results.

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -254,32 +232,43 @@ class WooPosVariationsViewModel @Inject constructor(
+                    is WooPosProductsDataSource.VariationsResult.Cached -> {
+                        // Handle cached variations
+                        if (result.data.isNotEmpty()) {
+                            _viewState.value = createContentState(result.data, productId)
```
Cached results are intermediate, meaning we're still waiting for Remote results, so if empty Cached result is received, we should keep showing PTR indicator as we're awaiting Remote results.

samiuelson committed:
Merge branch `trunk` and resolve conflicts

samiuelson committed:
Merge branch 'trunk' into woomob-1566-woo-poslocal-catalog-ptr-runs-incremental-sync-even-on

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/products/WooPosProductsDataSource.kt:
```diff
@@ -223,6 +240,68 @@ class WooPosProductsInDbDataSource @Inject constructor(
+        }
+    }.flowOn(Dispatchers.IO)
+
+    private suspend fun shouldUseIncrementalSync(): Boolean {
```
❓  I'm not sure I follow the code - why do we invoke `fetchFirstVariationsPage` from refreshVariations 😕 ?

samiuelson replied:
Fixed: 801786d76ccca43e62797219c19e43bb25635770

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/products/WooPosProductsDataSource.kt:
```diff
@@ -223,6 +240,68 @@ class WooPosProductsInDbDataSource @Inject constructor(
 
     override fun canLoadMoreVariations(numOfVariations: Int): Boolean = false
 
+    override suspend fun refreshProducts(): Flow<WooPosProductsDataSource.ProductsResult> = flow {
```
❓ It feels a bit off that `refresh` returns a flow, especially since it always emits just one response. Have you considered returning `Result<Unit>` and relying on Room updating `getProductsFromDatabaseFlow` that is being observed? Alternatively, we could return Result<ProductsResult> which isn't optimal but might simplify implementing the remote approach.

samiuelson replied:
@malinajirka, thanks for your question. This is not obvious how to marry the Hybrid and LocalCatalog incremental sync results. Technically, the Hybrid returns a flow of `Cached` (intermediate) and `Remote` (final) results. However, in case `forceRefresh=true`, the cache gets cleared, I believe we can be interested solely in the `last` emitted result and can return it as a result. Could you please take a look at my refactor commit here: a7fd7dbd322dd82de15d997d9e0124849e935a27 and let me know what you think about the approach of updating/consolidating the data sync result models?

malinajirka replied:
> This is not obvious how to marry the Hybrid and LocalCatalog incremental sync results. 

Yeah you are right, it's not trivial. I think ideally both approaches would return data from cache - the only difference should be whether it's persistent db or in-memory cache. The fact that we return the response directly isn't optimal, as it means we have multiple sources of truth. Having said that, I understand that's out of scope. Moreover, the VM logic currently determines whether to hide/show progress/error-screen, based on whether the returned data are cache/remote, so it's not something we could easily refactor.


> Could you please take a look at my refactor commit here: https://github.com/woocommerce/woocommerce-android/commit/a7fd7dbd322dd82de15d997d9e0124849e935a27 and let me know what you think about the approach of updating/consolidating the data sync result models?

I like it!

samiuelson committed:
Remove comment

samiuelson committed:
Rename function

samiuelson committed:
Fix implementation of `refreshProducts` and `refreshVariations` in `WooPosProductsInDbDataSource`

samiuelson committed:
Refactor: Unify sync result types and observe product/variation lists

This commit refactors the data synchronization and observation logic within the WooPOS items list.

Key changes:
- Introduced `WooPosSyncResult` as a common sealed interface for different synchronization results.
- Moved `ProductsResult` and `VariationsResult` into a new `WooPosHybridSyncResult` file, which inherits from `WooPosSyncResult`, to centralize result type definitions.
- Modified `refreshProducts` and `refreshVariations` in data sources and ViewModels to return `Result<WooPosSyncResult>`, simplifying the handling of refresh outcomes.
- Implemented continuous observation of products and variations in `WooPosProductsViewModel` and `WooPosVariationsViewModel` to ensure the UI immediately reflects data from the cache upon initialization.

samiuelson committed:
Merge branch 'trunk' into woomob-1566-woo-poslocal-catalog-ptr-runs-incremental-sync-even-on

samiuelson committed:
Remove unreachable code

samiuelson committed:
Remove redundant code

samiuelson committed:
Remove unreachable code

samiuelson committed:
Update tests

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/products/WooPosProductsDataSource.kt:
```diff
@@ -232,6 +236,24 @@ class WooPosProductsInDbDataSource @Inject constructor(
+        selectedSite.getOrNull()?.let { site ->
+            val syncResult = localCatalogSyncRepository.syncLocalCatalogIncremental(site)
+            Result.success(syncResult)
+        } ?: Result.success(
```
❓ Is this intentionally returning `Success` when site is not selected?

samiuelson replied:
Fixed: 699b021

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/products/WooPosProductsDataSource.kt:
```diff
@@ -232,6 +236,24 @@ class WooPosProductsInDbDataSource @Inject constructor(
+        selectedSite.getOrNull()?.let { site ->
+            val syncResult = localCatalogSyncRepository.syncLocalCatalogIncremental(site)
+            Result.success(syncResult)
+        } ?: Result.success(
```
Same as above.

samiuelson replied:
Fixed: 699b021

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/products/WooPosProductsDataSource.kt:
```diff
@@ -488,6 +510,24 @@ class WooPosProductsRemoteDataSource @Inject constructor(
             ?.toWooPosVariation(variationMapper)
     }
 
+    override suspend fun refreshProducts(): Result<WooPosSyncResult> = withContext(
```
> when forceRefresh = true the in-memory cache gets wiped out

I can't fine the convo right now, but Andrei mentioned we have not been wiping cache on PTR - I remember we used to do it early in the POS but then switched to keeping the in-memory cache until app restart. Do you know at which point we started calling `productsIndex.clearCache()` in the data source implementation? I mean, why we decided to do it?

samiuelson replied:
I just found out that the only place the in-memory cache is cleared is in `WooPosProductsDataSource::prepopulateProductsCache` which is called in `WooPosSplashViewModel`. On PTR only the productsIndex is cleared (responsible for pagination) — not the in-memory cache (`WooPosProductsCache`). This means that we are not clearing the cache and Andrey's issue is related to something else.

malinajirka replied:
Ahh, good find! So this PR is ready to be tested, right?

samiuelson replied:
Yes, ready 👍

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/variations/WooPosVariationsViewModel.kt:
```diff
@@ -254,32 +228,60 @@ class WooPosVariationsViewModel @Inject constructor(
+            val result = dataSource.refreshVariations(productId)
+            result.onSuccess { posSyncResult ->
+                when (posSyncResult) {
+                    is com.woocommerce.android.ui.woopos.localcatalog.PosLocalCatalogSyncResult.Success -> {
```
🔍 We could remove the full path.

samiuelson replied:
Done: fa6161a

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/localcatalog/WooPosHybridSyncResult.kt:
```diff
@@ -0,0 +1,16 @@
+import com.woocommerce.android.ui.woopos.common.data.WooPosVariation
+import com.woocommerce.android.ui.woopos.common.data.models.WooPosProductModel
+
+sealed class WooPosHybridSyncResult : WooPosSyncResult()
```
💡 I think I understand the idea here, but I would consider doing it the other way round. The current implementation returns a result which can be either ProductsResult or VariationsResults (for hybrid approach). That is a bit misleading and forces us to ignore variationsResult in product VM and product result in Variations VM. 

Instead, we could have two almost identical copies for local Catalog:
```
sealed class WooPosSyncProductResult
sealed class WooPosSyncVariationResult
```


Local:
```
sealed class PosLocalCatalogProductSyncResult : WooPosSyncProductResult() {
        val value: PosLocalCatalogSyncResult    
}

sealed class PosLocalCatalogSyncResult : WooPosSyncVariationResult() {
    val value: PosLocalCatalogSyncResult
}


sealed class PosLocalCatalogSyncResult {
    data class Success(
        val productsSynced: Int,
        val variationsSynced: Int,
        val syncDurationMs: Long
    ) : PosLocalCatalogSyncResult()

    sealed class Failure(val error: String) : PosLocalCatalogSyncResult() {
        class CatalogTooLarge(error: String, val totalPages: Int, val maxPages: Int) : Failure(error)
        class UnexpectedError(error: String) : Failure(error)
    }
}
```

Hybrid:
```
sealed class WooPosHybridProductSyncResult : WooPosSyncProductResult()
sealed class WooPosHybridVariationSyncResult : WooPosSyncVariationResult()

```

Wdyt?

malinajirka replied:
No need to update this - the more I think about it the more I feel like perhaps the downside of the current approach aren't that significant to justify this refactoring.

samiuelson replied:
I like the idea, done: 5cefb66

malinajirka reviewed (commented):
Thanks @samiuelson !

As we discussed I haven't tested this PR since I'll test it when we update the behavior of cleaning cache on PTR.

I've left some suggestions/questions but it looks good overall and I believe can be merged as is.

malinajirka reviewed (approved):
Works as expected!

samiuelson committed:
Clean up code

samiuelson committed:
Return failure if site is not selected

samiuelson committed:
Introduce specific sync result types for products and variations

This commit refactors the synchronization result handling by introducing more specific sealed classes. `WooPosSyncProductResult` and `WooPosSyncVariationResult` are created to distinguish between product and variation sync operations.

This change eliminates the need for `WooPosHybridSyncResult` and prevents unexpected result types in view models, improving type safety and code clarity.

samiuelson committed:
Merge branch 'trunk' into woomob-1566-woo-poslocal-catalog-ptr-runs-incremental-sync-even-on

samiuelson committed:
Refactor WooPosSyncResult to remove unused sealed class

samiuelson committed:
Satisfy detekt's complaints
